### PR TITLE
resolve issue with missing template error

### PR DESCRIPTION
**Pull request by brad-decker**
Fixes: [METAMASK-GSGN](https://sentry.io/organizations/metamask/issues/2262591771/?environment=production&project=273505&query=wallet_requestPermissions&statsPeriod=14d)

Possible cause?
```js
    this.setState({ mounted: true });
    if (isNotification && totalUnapprovedCount === 0) {
      global.platform.closeCurrentWindow();
    } else if (!isNotification && showAwaitingSwapScreen) {
      history.push(AWAITING_SWAP_ROUTE);
    } else if (!isNotification && haveSwapsQuotes) {
      history.push(VIEW_QUOTE_ROUTE);
    } else if (!isNotification && swapsFetchParams) {
      history.push(BUILD_QUOTE_ROUTE);
    } else if (firstPermissionsRequestId) { // <------ If this is undefined, which is an approval
      history.push(`${CONNECT_ROUTE}/${firstPermissionsRequestId}`);
    } else if (unconfirmedTransactionsCount > 0) {
      history.push(CONFIRM_TRANSACTION_ROUTE);
    } else if (Object.keys(suggestedTokens).length > 0) {
      history.push(CONFIRM_ADD_SUGGESTED_TOKEN_ROUTE);
    } else if (pendingApprovals.length > 0) { // <------- Then this would be true
      history.push(CONFIRMATION_V_NEXT_ROUTE);
    }
```

Why would firstPermissionsRequestId be undefined when there is pending approval of this type? Not sure. This change just makes it so that the only approvals that will be caught on the last block of the else/if are ones that have templates. This might mean for that user that they don't get routed to the notification at all -- perhaps not a real solution but one that doesn't cause an error.

### CODE REVIEW

**Review by Welziewagers2015 — DISMISSED** [review]

**Comment by Gudahtt** [comment]
>Why would firstPermissionsRequestId be undefined when there is pending approval of this type? Not sure.

I think this is the question we need to answer. I don't have much confidence that this change would make things less-broken without better understanding what went wrong here.

**Comment by Gudahtt** [comment]
I've just traced through how these permission requests are handled, and it looks like we store two different related states for each request. We have `permissionsRequests` which tracks a request for permissions, and we have the Approval controller state which is the user confirmation asking for approval for this permission request.

It looks like the lifetime of the `permissionsRequests` entry should _exceed_ the lifetime of the approval request :thinking: At least that's what I'm seeing. Either when rejecting or approving the confirmation, the approval request is removed first, before the permissions request. There must be some edge case at play here...

**Comment by brad-decker** [comment]
I think this change is still likely a good one, but should be batched together with any adjustments we want to make to this flow to be safe. I'll also look through the trace and see if I can pinpoint a culprit or suggest a change to make this safer.

**Comment by Gudahtt** [comment]
I agree that we should protect against routing things without templates to the template-based confirmation page. But I'm concerned that this PR as-written might result in the error being hidden, which would make fixing it properly more difficult.

**Comment by brad-decker** [comment]
>I agree that we should protect against routing things without templates to the template-based confirmation page. But I'm concerned that this PR as-written might result in the error being hidden, which would make fixing it properly more difficult.

I agree, that's what I meant by batching together, but in retrospect, that wasn't clearly decipherable from what I wrote.

**Comment by brad-decker** [comment]
If we rewind to before the template system/approval part of this logic, there wasn't an else statement catching errors there. I just added another `else if` onto the chain. Unless something changed with the permissions/approval controllers that landed alongside the template system, not before it, then the edge case existed then. We didn't know about it because it doesn't present as an error.

So I'd be in favor of adding an else statement, but maybe instead of throwing an Error we should just log it to sentry? Doesn't throwing an error in that case cause the user flow to be interrupted?

**Comment by Gudahtt** [comment]
>So I'd be in favor of adding an else statement, but maybe instead of throwing an Error we should just log it to sentry? Doesn't throwing an error in that case cause the user flow to be interrupted?

Sounds reasonable in principle, but what would happen to the user in this case? Throwing an error might be all we can do with the information we have, if it's an unaccounted for situation.

**Comment by brad-decker** [comment]
> Sounds reasonable in principle, but what would happen to the user in this case? Throwing an error might be all we can do with the information we have, if it's an unaccounted for situation.

Yeah, fair. In any event, we should think about what to do in this case once *all* confirmations are using the approval controller. In that case, if we get to the point where the confirmation is not in the template system/not cared for in our logic, we could automatically reject it on the user's behalf and display a warning to the user while recording the details of the approval. That'd give us enough clues to work with I think.

File: ui/app/pages/confirmation/confirmation.js
Comment (Gudahtt) [inline ui/app/pages/confirmation/confirmation.js:118]: Was this the underlying error? I'm not quite sure I follow :thinking: 

My guess is that the problem occurs when there is a "request permissions" confirmation queued up behind a templated confirmation. I don't yet understand why it only affects the request permissions confirmation though :thinking:

Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js:118]: `wallet_requestPermissions` uses the approval controller, which creates an entry in the `pendingApprovals` object. I was expecting that type of approval to always be handled by the routing login in the home component, and supersede this new page ever being called. However, if a `wallet_addEtherumChain` is queued up *before* a `wallet_requestPermissions`, the routing logic is skipped because there is already a notification window open and the user is viewing the `addEthereumChain` confirmation. Once they click confirm this page would have attempted to pull the remaining approval out which would be the `wallet_requestPermissions` one.

Comment (Gudahtt) [reply ui/app/pages/confirmation/confirmation.js:118]: Ahhh I see! That makes sense, thanks

File: ui/app/pages/home/home.container.js
Comment (Gudahtt) [inline ui/app/pages/home/home.container.js:106]: Nit: it might be worth renaming this prop, now that it no longer includes all pending approvals. Maybe something like `pendingConfirmations`, like you used on the confirmation page.

**Review by Gudahtt — DISMISSED** [review]
LGTM!

It would be better if we handled the `else` case, but, not a blocker as that's a pre-existing problem.

**Comment by brad-decker** [comment]
@Gudahtt gonna do the else block in a different PR -- need to think through the logic there. it won't be an else block, but rather another else-if or perhaps reworking the entire structure of this logic because the else block will trigger always because its in the `componentDidMount` method.

**Review by Gudahtt — APPROVED** [review]

**Pull request merged**
